### ISSUE TITLE
Poor performance and warning for GradientTape.jacobian with experimental_use_pfor=False

### ISSUE THREAD

**Issue by cmichelenstrofer** [tensorflow/tensorflow#36227]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Scientific Linux 7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): conda
- TensorFlow version (use command below): 2
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Extremely slow gradient calculation. 
Also the following warning: 
```
WARNING:tensorflow:From /home/grad3/cmich/miniconda3/envs/3.7/lib/python3.7/site-packages/tensorflow_core/python/ops/array_grad.py:502: _EagerTensorBase.cpu (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.identity instead.
```

**Describe the expected behavior**
There most be a much faster way of getting these gradients. After all the TensorFlow optimization must use these and those evaluations do not take so long. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```python
import time
import tensorflow as tf
import numpy as np

class Model(tf.keras.Model):
    def __init__(self,):
        super(Model, self).__init__()
        self.layers_list = [tf.keras.layers.Dense(2)] + \
            [tf.keras.layers.Dense(1000)] + \
            [tf.keras.layers.Dense(500)]

    def call(self, x):
        for layer in self.layers_list:
            x = layer(x)
        return x

model = Model()
model.build((1,2))
weights = model.trainable_variables

input = np.zeros([1,2])

with tf.GradientTape(persistent=True) as tape:
    ts = time.time()
    output = model(input)
    print(f'Forward took {time.time()-ts:.2f}s')

ts = time.time()
gradients = tape.jacobian(output, weights, experimental_use_pfor=False)
print(f'Backward took {time.time()-ts:.2f}s')
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
Output:
```
Forward took 0.01s
WARNING:tensorflow:From /home/grad3/cmich/miniconda3/envs/3.7/lib/python3.7/site-packages/tensorflow_core/python/ops/array_grad.py:502: _EagerTensorBase.cpu (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.identity instead.
Backward took 5.75s
```
The gradient calculation takes several orders of magnitude more time (for my real case the respective times are about 0.03s and 18s which is too much). 

Is this the best way of doing this or this a faster way? Also should I care about the warning?

**Comment by byronyi** [comment]
Any reasons why you are setting `experimental_use_pfor=False`? `pfor` means "parallel for" and it brings greater parallelism.

**Comment by cmichelenstrofer** [comment]
@byronyi Indeed I had to use both the ``experimental_use_pfor=False`` and the ``persistent=True`` options for my real case (would get an error and a message saying I needed to). But I did not check if this toy case would run without them. I will check in more detail tomorrow when I am back in front of a computer and update the toy case if necessary to better reflect my real case.

**Comment by cmichelenstrofer** [comment]
@byronyi I want to use LeakyReLU activation. When I add this layer then I get the following error:
```
Encountered an exception while vectorizing the jacobian computation. Vectorization can be disabled by setting experimental_use_pfor to False.
```


The modified toy case is:
```
import time
import tensorflow as tf
import numpy as np

class Model(tf.keras.Model):
    def __init__(self,):
        super(Model, self).__init__()
        self.layers_list = [tf.keras.layers.Dense(2)] + \
            [tf.keras.layers.Dense(1000)] + \
            [tf.keras.layers.LeakyReLU(alpha=0.1)] + \
            [tf.keras.layers.Dense(500)]

    def call(self, x):
        for layer in self.layers_list:
            x = layer(x)
        return x

model = Model()
model.build((1,2))
weights = model.trainable_variables

input = np.zeros([1,2], dtype='float32')

# with tf.GradientTape(persistent=True) as tape:
with tf.GradientTape() as tape:
    ts = time.time()
    output = model(input)
    print(f'Forward took {time.time()-ts:.2f}s')

ts = time.time()
# gradients = tape.jacobian(output, weights, experimental_use_pfor=False)
gradients = tape.jacobian(output, weights)
print(f'Backward took {time.time()-ts:.2f}s')
```

**Comment by cmichelenstrofer** [comment]
Also, with the original toy case (without LeakyReLU) if I don't use of the ``experimental_use_pfor=False`` I get the following:

```
Forward took 0.01s
2020-01-27 10:35:40.909307: E tensorflow/core/common_runtime/executor.cc:642] Executor failed to create kernel. Internal: No function library
	 [[{{node loop_body/MatMul_4/pfor/cond}}]]
2020-01-27 10:35:40.950468: E tensorflow/core/common_runtime/executor.cc:642] Executor failed to create kernel. Internal: No function library
	 [[{{node loop_body/MatMul_3/pfor/cond}}]]
2020-01-27 10:35:40.991867: E tensorflow/core/common_runtime/executor.cc:642] Executor failed to create kernel. Internal: No function library
	 [[{{node loop_body/MatMul_1/pfor/cond}}]]
Backward took 1.03s
```

**Comment by agarwal-ashish** [comment]
It looks like you are computing jacobian on an output of size 500. Without pfor, it would translate to 500 evaluations of gradient, so a slowdown of 500x is as expected. It looks like pfor vectorization provides a 5x speedup on top of an iterative strategy which also sounds in line with expectation. So I'd say that this is working as intended.

As an aside, If you plan on using a batch size larger than 1, I'd recommend batch_jacobian instead of jacobian.

**Comment by gowthamkpr** [comment]
@cmichelenstrofer Please let me know if the issue has been resolved using the above comment. Thanks!

**Comment by cmichelenstrofer** [comment]
If this is how slow it is supposed to be than I guess there's nothing to do. Unfortunately, I cannot use batch normalization for my case. I was just surprised it was so slow when the optimization/training in tensorflow is not this slow and it requires gradient calculation. But from what I've been reading it seems tensorflow never calculates the full jacobian when doing back-propagation/gradient descent.

**Comment by gowthamkpr** [comment]
@agarwal-ashish Can you PTAL? Thanks!

**Comment by agarwal-ashish** [comment]
@cmichelenstrofer can you clarify the requirement ? Typically networks have a scalar loss at the top which makes the jacobian computation cheap. For non-scalar output, batch jacobian would be faster if you can use it (batch entries are independently transformed). If you compute full jacobian with non-scalar outputs, then yes, it will be exepected to be slow.

**Comment by cmichelenstrofer** [comment]
@agarwal-ashish The requirement is for non-scalar output and no batches. Very high-dimensional input and output spaces. And I need the Jacobian of all outputs w.r.t. all inputs (no batch Jacobian). I was under the impression that the full Jacobian needed to be computed as an intermediate step even when using a scalar output (a scalar output is after all one extra step in the forward model). This seems to not be the case and you use a numerical trick to avoid computing the large intermediate Jacobian (is this correct?). So counter-intuitively adding an extra step to your forward model makes the back-propagation cheaper. If this is indeed the case, it sounds like there is no way around the jacobian being so expensive.

**Comment by agarwal-ashish** [comment]
Gradient functions are implemented as vector-jacobian products, and they often use tricks for more efficient computation which avoids creating jacobian matrices.
GradientTape.Jacobian still builds upon those tricks, but effectively it is now running many of the vector-jacobian computations, which it vectorizes for 5-10x speedups.
